handle postgres cancel requests

GavinFrazar opened the pull request:
Closes #9389

This PR adds support for postgres cancel requests.

smallinsky reviewed (commented):
First pass.

greedy52 reviewed (commented):
First pass

FYI, seems RDS Proxy doesn't support cancel request XD so we are doing better
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-proxy.html

GavinFrazar commented on lib/srv/db/postgres/engine.go:
```diff
@@ -107,6 +113,10 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 	}
+	if e.cancelReq != nil {
+		// Special case when sending a cancel request.
+		return e.handleCancelRequest(ctx, sessionCtx)
```
no audit log event. Since cancel requests establish a new connection, the only way to associate a cancel request with a session is by adding PID to the session as well, and there's no way to know if a cancel request actually cancelled anything either. And then every time the user hits `ctrl+c`, there will be audit log events for a new session start, cancel request sent, session end. I feel like there were other problems I'm forgetting that came up when discussing this with @smallinsky last week, but we opted to skip audit logging changes at least for now.

GavinFrazar commented on tool/tsh/db.go:
```diff
@@ -641,47 +643,43 @@ func maybeStartLocalProxy(ctx context.Context, cf *CLIConf,
+	route            tlsca.RouteToDatabase
+	database         types.Database
+	listener         net.Listener
+	autoReissueCerts bool
+	tunnel           bool
```
tunnel mode with auto reissue certs happens when you use `tsh proxy db --key-file=... --cert-file=...`. I'll add a comment

GavinFrazar commented on lib/utils/certs.go:
```diff
@@ -286,8 +286,13 @@ func NewCertPoolFromPath(path string) (*x509.CertPool, error) {
+// from a tls.Certificate.
+// If the leaf certificate is not parsed already, then this function parses it.
+func TLSCertLeaf(cert tls.Certificate) (*x509.Certificate, error) {
+	if cert.Leaf != nil {
+		return cert.Leaf, nil
+	}
```
handshake never sets this field, so this is safe.

```
	// Leaf is the parsed form of the leaf certificate, which may be initialized
	// using x509.ParseCertificate to reduce per-handshake processing. If nil,
	// the leaf certificate will be parsed as needed.
```

GavinFrazar commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -205,14 +230,27 @@ func (l *LocalProxy) GetAddr() string {
 func (l *LocalProxy) handleDownstreamConnection(ctx context.Context, downstreamConn net.Conn) error {
 	defer downstreamConn.Close()
 
+	var certs []tls.Certificate
+	shouldTunnel, downstreamConn, err := l.shouldTunnelConn(ctx, downstreamConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if shouldTunnel {
+		l.cfg.Log.Debug("Loading client certs to dial remote proxy.")
+		certs, err = l.getCerts()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
```
prior to this change, the only thing that determined if certs were used to dial upstream (i.e. like `--tunnel`) was whether certs were set in the local proxy. Since we wanted to support non-tunnel postgres local proxy `tsh proxy db postgres`, I had to load certs from the profile in case tunnel was required for a cancel request. But then certs were loaded into local proxy but should only be used for cancel requests. So yes it's needed due to that requirement

GavinFrazar commented on lib/srv/db/postgres/test.go:
```diff
@@ -306,3 +420,10 @@ var TestQueryResponse = &pgconn.Result{
+
+// TestLongRunningQuery is a stub SQL query clients can use to simulate a long
+// running query that can be only be stopped by a cancel request.
+const TestLongRunningQuery = "pg_sleep(forever)"
+
+// TestSecretKey is the secret key stub for all connections, used for cancel requests.
+const TestSecretKey = 1234
```
no, I'll unexport

GavinFrazar replied:
oh, actually yes we need to export `TestLongRunningQuery`. I'll just unexport `TestSecretKey`

GavinFrazar commented on lib/srv/db/postgres/proxy.go:
```diff
@@ -141,6 +144,7 @@ func (p *Proxy) handleStartup(ctx context.Context, clientConn net.Conn) (*pgprot
 		// https://www.postgresql.org/docs/13/protocol-flow.html#id-1.10.5.7.11
 		switch m := startupMessage.(type) {
 		case *pgproto3.SSLRequest:
+			p.Log.Debugf("Received startup message: %#v.", startupMessage)
```
I just didn't want to log the cancel request with `%#v` since that would log the secret key.

GavinFrazar replied:
I'll add a patch that handles cancel request separately to remove the duplication.

GavinFrazar commented on lib/srv/db/postgres/engine.go:
```diff
@@ -107,6 +113,10 @@ func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Sessio
 	}
+	if e.cancelReq != nil {
+		// Special case when sending a cancel request.
+		return e.handleCancelRequest(ctx, sessionCtx)
```
sure

GavinFrazar commented on lib/srv/db/postgres/engine.go:
```diff
@@ -459,6 +473,61 @@ func (e *Engine) getConnectConfig(ctx context.Context, sessionCtx *common.Sessio
+	if response[0] != 'S' {
+		return nil, trace.Errorf("server refused TLS connection")
+	}
+	return tls.Client(conn, tlsConfig), nil
```
There is no need to `trace.Wrap` trace errors.

GavinFrazar commented on tool/tsh/db.go:
```diff
@@ -1144,6 +1170,10 @@ func withConnectRequirements(ctx context.Context, tc *client.TeleportClient, rou
 			// via `tsh db connect`.
 			// (elasticsearch-sql-cli cannot be configured to use specific certs).
 			r.addLocalProxyWithTunnel(formatDBProtocolReason(route.Protocol))
+		case defaults.ProtocolPostgres:
+			// postgres needs a local proxy to support cancel requests.
+			r.addLocalProxy(fmt.Sprintf("%v and cancel requests require local proxy",
+				formatDBProtocolReason(route.Protocol)))
```
note that this is in `withConnectRequirements`, as in `tsh db connect`. So this will only ensure we start  a local proxy with `tsh db connect`.

GavinFrazar replied:
is that ok?

smallinsky replied:
> So this will only ensure we start a local proxy with tsh db connect.
Even if this is only `tsh db connect` the default behaviour was changed. If a customer was using Separate listener mode the `tsh db connect`  flow was not using local proxy. After that change if cluster was configured with separate_listener_mode  the `tsh db connect` will trigger TLS Routing flow. 
My main point is not to change this behaviour.

GavinFrazar replied:
Is the concern here due to local proxy causing TLS-inside-TLS behavior? Because without this change, `tsh db connect` will not support cancel requests in separate listener mode.

smallinsky replied:
> Is the concern here due to local proxy causing TLS-inside-TLS behavior? 
I don't that that TLS inside TLS is  our concern. We are using this behavior in other integration but we don't want to change/alter suddenly current behavior and make the `--tunnel` mode as default for all postgres connection. 
  
Why the proxy cannot run without --tunnel mode for normal connection and upgrade a new connection to TLS only if the cancel request was detected ?

smallinsky replied:
@GavinFrazar 
Even if we upgrade the connection only to TLS without `--tunnel` option this behaviour can break the client integration. 

Let's imagine that the client is using the following sperate teleport proxy configuration: 
```
auth_service:
  proxy_listener_mode: "separate"
  
proxy_service:
  enabled: "true"
  public_addr:  ['alb.master.ice-berg.dev:3080']
  ssh_public_addr: ['master.ice-berg.dev:3023']
  tunnel_public_addr: ['master.ice-berg.dev:3024']
  postgres_public_addr: ['master.ice-berg.dev:5432']
  web_listen_addr: 0.0.0.0:3080
  postgres_listen_addr: 0.0.0.0:5432
```

where `alb.master.ice-berg.dev:`   is an L7 Load Balancer like AWS ALB that terminates TLS connection and dials the teleport proxy without forwarding TLS extensions. 

before this change the `tsh db connect postgres` command was using the `master.ice-berg.dev:3024` address after switching to a local proxy for postgres connection the `tsh db connect postgres` will try to connect to `alb.master.ice-berg.dev:3080` with TLS Routing extension. Due to L7 ALB the TLS exertion will be dropped and the client will not be able to connect to the right service. Recently this case was fixed by adding support for connection upgrade but I can still imagine that the v13 client can be used to connect to the old teleport proxy that doesn't support [Connection Upgrade](https://github.com/gravitational/teleport/issues/14369) feature.

greedy52 replied:
Hey guys. I know we have code to track what mode is required, when, which protocol, etc.. but I've decided to create a spreadsheet to track it as it's slightly more visual. Feel free to update it or fix it if something is not accurate:
https://docs.google.com/spreadsheets/d/1m0swp5F4PZ9-C6pAUR45FbIB1o2IeJ3mjmJz1VrRG3g/edit?usp=sharing

GavinFrazar commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -360,3 +408,59 @@ func (l *LocalProxy) SetCerts(certs []tls.Certificate) {
+		// So when the local proxy has no certs configured, we inspect
+		// the connection to see if it is a postgres cancel request and
+		// load certs for the connection.
+		startupMessage, conn, err := peekPostgresStartupMessage(ctx, downstreamConn)
```
It's possible and I considered doing something like that, but it would require a lot more refactoring and complexity to do it. FWIW, all of the local proxy changes are due to the requirement that `tsh proxy db` (without `--tunnel` or `--cert-file/--key-file`) loads and uses certs for cancel requests, aka "tunnels on demand". If we instead just let cancel requests fail when client certs are not presented, then all of this extra complexity goes away. I consider it a bug that `psql` doesn't send cancel requests with SSL when I've configured `psql` to connect with client certs.

greedy52 replied:
Ya, refactoring can get complicated. The easiest i can think of is to let `middleware.OnNewConnection` return a per connection setting for overwriting downstreamConn and TLS certs. Still a lot of work to change existing middleware but the benefit would be to encapsulate cancel request stuff in a separate middleware that is only used when connecting to Postgres in non-tunnel mode. I'll leave it to you to decide if you are in the mood to refactor in some way or keep it as it is.

regardless, it would be great if we can move pgproto3 stuff to a separate util lib.

GavinFrazar commented on integration/appaccess/pack.go:
```diff
@@ -385,6 +385,7 @@ func (p *Pack) startLocalProxy(t *testing.T, tlsConfig *tls.Config) string {
 		Listener:           listener,
 		ParentContext:      context.Background(),
 		Certs:              tlsConfig.Certificates,
+		Tunnel:             true,
```
`CheckCertsNeeded` flag instead? I'm not sure what to name it. This comes from the requirement that `tsh proxy db` tunnels only as needed, whereas normally that command would not load and use client certs ever. If we instead allowed cancel requests to fail when not using `--tunnel`, then this complexity goes away. In my opinion, it is a client software error to send cancel requests without SSL when configured to use SSL (like `psql` does).

greedy52 commented on lib/srv/db/postgres/proxy.go:
```diff
@@ -185,15 +201,12 @@ func (p *Proxy) handleStartup(ctx context.Context, clientConn net.Conn) (*pgprot
-			case *alpnproxy.PingConn:
-				return m, tlsConn, backend, nil
-			default:
+			tlsConn, ok := clientConn.(utils.TLSConn)
+			if !ok {
```
good catch

greedy52 commented on lib/srv/db/postgres/test.go:
```diff
@@ -297,7 +365,53 @@ func (s *TestServer) ParametersCh() chan map[string]string {
+func (s *TestServer) cleanupPid(pid uint32) {
+	s.pidMu.Lock()
+	defer s.pidMu.Unlock()
+	if entry, ok := s.pids[pid]; ok && entry != nil && entry.cancel != nil {
+		entry.cancel()
+		delete(s.pids, pid)
```
nit: should we delete even if cancel is nil?

GavinFrazar replied:
I think we might as well, although it doesn't really matter since PIDs are monotonically increasing to keep it simpler. I mainly wanted to prevent context leaks from failing to call cancel

greedy52 reviewed (commented):
LTGM overall except the one Marek raised about `tsh db connect` usage. I agree that we shouldn't use local proxy if there is a separate port for Postgres.

Personally I would still prefer abstracting the postgres specific logic away from the local proxy, somehow, if possible. We added a few bool parameter a few places but that didn't eliminate checking against Postgres protocol in each level. Doesn't have to be now so maybe we can think about it a localproxy/middleware refactoring/redesign separately in the future to make this better

GavinFrazar commented:
@smallinsky @greedy52 updated to move the postgres local proxy requirement to only when TLS routing is enabled per marek's suggestion, please take another look

GavinFrazar commented:
This local proxy requirement for postgres with TLS routing mode is wrong I think.

Local proxy postgres with TLS routing should only require a local proxy when the proxy is behind an L7 load balancer.

GavinFrazar commented:
@smallinsky `tsh db connect` will always use a local proxy when TLS routing is enabled. That's here:
```
func withConnectRequirements(ctx context.Context, tc *client.TeleportClient, route *tlsca.RouteToDatabase) requireOpt {
	return func(r *dbLocalProxyRequirement) {
		if !r.localProxy && tc.TLSRoutingEnabled {
			r.addLocalProxy(formatTLSRoutingReason(tc.SiteName))
		}
```

What I'm saying is that adding a requirement for all the other commands in `getDBLocalProxyRequirement` is wrong. As you mention, users can connect directly without local proxy, so `tsh db config` and `tsh db env` should still for postgres with TLS routing enabled.

So I think I should revert the change in `getDBLocalProxyRequirement` here:
https://github.com/gravitational/teleport/pull/22173/commits/dd2f5f5ad64370cd6cd62dd98bd7ada80b89f739#diff-dae87d4d3799335d85a5fa6ff17cfcbfb30705ae145c9e67ef70286423676a48R1153-R1154

GavinFrazar commented:
The broader issues with L7 load balancer and other db protocols need to be addressed in another PR. We should probably make use of `IsALPNConnUpgradeRequired` (and maybe save that info in the user's .tsh profile) for that purpose.

I think this PR is otherwise ready to merge, if you approve of removing local proxy requirement for postgres with TLS routing in `getDBLocalProxyRequirement` then we can merge this.
```diff
	case defaults.ProtocolMySQL,
-		defaults.ProtocolPostgres:
		if tc.TLSRoutingEnabled {
			out.addLocalProxy(fmt.Sprintf("%v and %v",
				formatDBProtocolReason(route.Protocol),
				formatTLSRoutingReason(tc.SiteName)))
		}
```

greedy52 commented:
> The broader issues with L7 load balancer and other db protocols need to be addressed in another PR. We should probably make use of IsALPNConnUpgradeRequired (and maybe save that info in the user's .tsh profile) for that purpose.

I will work on this next week to address it separately. And agreed we shouldn't require local proxy for Postgres for `tsh db env/config` when TLS routing is enabled.

GavinFrazar committed:
handle postgres cancel requests

* support cancel requests in proxy and engine
* add cert leaf init util func
* init local proxy cert leaves
* inspect non-tunnel local proxy connections
  - check for postgres cancel requests
  - tunnel postgres cancel requests "on-demand"
